crates/shared/src/order_validation.rs

```diff
@@ -345,20 +345,14 @@ impl OrderValidator {
         self
     }
 
-    async fn check_max_limit_orders(
-        &self,
-        owner: H160,
-        class: &OrderClass,
-    ) -> Result<(), ValidationError> {
-        if class.is_limit() {
-            let num_limit_orders = self
-                .limit_order_counter
-                .count(owner)
-                .await
-                .map_err(ValidationError::Other)?;
-            if num_limit_orders >= self.max_limit_orders_per_user {
-                return Err(ValidationError::TooManyLimitOrders);
-            }
+    async fn check_max_limit_orders(&self, owner: H160) -> Result<(), ValidationError> {
+        let num_limit_orders = self
```

Comment:
This will still consider in market orders that have previously been placed as limit orders and thus not allow you to place another limit order in case your total number of orders are above a threshold.

Probably not a huge deal, but seeing this now, I think we might actually have to change the db query that fetches limit orders to consider their quotes.

Reply:
@fleupold @MartinquaXD  Yes, but that is legacy behavior. We could still have internally `Market` and `Limit`, it would make it easier, WDYT?
Fetching the quotes and then matching it in order to determine if it is in market order may be not very efficient, because we need to add extra query call to the database, and then match the order.

We could do (in several steps):
- make an internal class with { `Market`, `Limit`}
- remove the `class` from the API (AFAIK, we want to remove `Liquidity` and `Market`)
- based on the in market value determine if the internal class is `Market` or `Limit`
- store it in the `class` column in the database
- create a DB migration to change all the `Liquidity` orders to `Limit` orders
- count the order limits only with `Limit` orders

This way we can:
- be sure we classify the orders accordingly, and we count them well
- enable future/potential statistics of the type of the order the users do (as it is right now it is very costly to check the type of orders)
- potentially add more classes if needed
- potentially limit the number of in market orders
- have full control of both types of orders

What do you think? :thinking:

Reply:
>  Yes, but that is legacy behavior.

True but also by now this is a bug and could be changed.

> Fetching the quotes and then matching it in order to determine if it is in market order may be not very efficient

Quotes already have an indexed `order_uid` so we can have fast joins for that purpose.

Improving correctness around order classes in the system sounds nice, but I would first measure that computing "in market-ness" is actually significantly slower because that seems to have guided this plan.

Reply:
> > Yes, but that is legacy behavior.
> 
> True but also by now this is a bug and could be changed.
Yes, definitely! What I meant about the legacy behavior was that I wasn't 100 % sure about the desired behavior in this case, so I left the legacy one when it was possible. 

> > Fetching the quotes and then matching it in order to determine if it is in market order may be not very efficient
> 
> Quotes already have an indexed `order_uid` so we can have fast joins for that purpose.
> 
> Improving correctness around order classes in the system sounds nice, but I would first measure that computing "in market-ness" is actually significantly slower because that seems to have guided this plan.

Clear!